File: CPP/ByTier/E/1862_E_Kolya_and_Movie_Theatre.cpp
// Problem: CF 1862 E - Kolya and Movie Theatre
// https://codeforces.com/contest/1862/problem/E

/*
E. Kolya and Movie Theatre

Algorithm: Sliding window with multiset to maintain the best m movies visited so far.
Key technique: Use a sliding window approach with a multiset to efficiently track the maximum sum of 
               entertainment values while ensuring at most m movies are selected.

Time Complexity: O(n log m) per test case, where n is the number of days and m is the maximum number of movies Kolya can attend.
Space Complexity: O(m) for storing the multiset.

The approach uses a greedy method with a sliding window:
- For each day i, we consider whether including the movie on that day is beneficial.
- We maintain a multiset of the last m entertainment values (in descending order).
- When adding a new movie, if the size exceeds m, we remove the smallest value to keep only the top m.
- The score for a visit on day i is calculated as: sum of selected movie values minus d * number of days since last visit.
- This is done by tracking the cumulative sum and adjusting for the decreasing value effect over time.

*/

#include <cstdio>
#include <set>
typedef long long ll;

int main() {
  ll t;
  scanf("%lld", &t);
  while (t--) {
    ll n, m, d;
    scanf("%lld %lld %lld", &n, &m, &d);
    std::multiset<ll> w; // multiset to maintain the top m entertainment values seen so far
    ll cur(0), mx(0);    // cur: current sum of selected movies, mx: maximum score found so far

    for (ll p = 1; p <= n; p++) {
      ll x;
      scanf("%lld", &x);
      if (x < 0) {
        continue; // skip non-positive entertainment values
      }

      cur += x;           // add current movie's value to the running sum
      w.insert(x);        // insert into multiset

      // if we've selected more than m movies, remove the smallest one
      if (w.size() > m) {
        cur -= *w.begin();  // subtract the smallest value from the sum
        w.erase(w.begin()); // remove it from the set
      }

      // calculate score for visiting on day p
      // total = (sum of selected movies) - (days since last visit) * d
      ll score = cur - p * d;
      mx = (mx > score) ? mx : score; // update max score
    }
    printf("%lld\n", mx);
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/1862/1862_E_Kolya_and_Movie_Theatre.py
# Problem: CF 1862 E - Kolya and Movie Theatre
# https://codeforces.com/contest/1862/problem/E

"""
Algorithm: Greedy with PriorityQueue (Min-Heap)
Approach:
- We process movies in order from day 1 to n.
- For each movie, if its entertainment value a[i] is positive, we consider including it.
- We maintain a min-heap of size at most m to keep track of the m most valuable movies we've seen so far.
- As we iterate, we keep track of total entertainment value (rsum) and update the maximum possible value of 
  (rsum - (idx + 1) * d) which represents total entertainment minus the penalty due to delay.

Time Complexity: O(n * log m) per test case, where n is number of movies and m is max visits allowed.
Space Complexity: O(m) for the heap storage.
"""

from heapq import heappush, heappushpop


def run():
    n, m, d = list(map(int, input().split()))
    a = list(map(int, input().split()))
    rsum = 0
    c = 0
    h = []
    ret = 0
    for idx, x in enumerate(a):
        if x < 0:
            continue  # Skip negative entertainment values
        rsum += x  # Add current movie's value to running sum
        if c < m:  # If we haven't reached the limit yet
            c += 1
            heappush(h, x)  # Add to heap
        else:
            t = heappushpop(h, x)  # Replace smallest in heap with current value
            rsum -= t  # Adjust sum accordingly
        tmp = rsum - (idx + 1) * d  # Calculate total entertainment with day count penalty
        if tmp > ret:
            ret = tmp  # Update maximum entertainment value
    print(ret)


TT = int(input())
for _ in range(TT):
    run()


# https://github.com/VaHiX/CodeForces/